pkg/admission/validator.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	v1alpha1 "sigs.k8s.io/service-apis/apis/v1alpha1"
+)
+
+// ValidateHTTPRoute checks if a httpRoute contains only 1 of
+// each HTTPRouteFilterType.
+func ValidateHTTPRoute(httpRoute v1alpha1.HTTPRoute) (bool, string, error) {
+	if len(httpRoute.Spec.Rules) == 0 {
+		return true, "", nil
+	}
+	numReqHeaderMod := 0
+	numReqMirror := 0
+	numExtRef := 0
+	for _, rule := range httpRoute.Spec.Rules {
+		if len(rule.Filters) != 0 {
+			for _, filter := range rule.Filters {
+				switch filter.Type {
+				case v1alpha1.HTTPRouteFilterRequestHeaderModifier:
+					numReqHeaderMod++
+				case v1alpha1.HTTPRouteFilterRequestMirror:
+					numReqMirror++
+				case v1alpha1.HTTPRouteFilterExtensionRef:
+					numExtRef++
+				}
```

Comment:
The code in this form has a higher maintenance burden because every time we add a new filter type, we need to ensure that it gets added here. Did you consider using a a map of string to int (or bool) to track the filters in the list and error out when you see multiple of those?

Reply:
This is a pattern that is used in the network policy API rules so I figured it would be a good pattern here too. 

I did not consider the map to int. I usually sway toward being explicit when working on API pieces, but I'm willing to change to the map if we think it's best. Would there be a perf overhead for this conversion?

Reply:
I like @hbagdi's idea for using a map here. I think we're likely going to be adding more of these types in the near future and it would be nice if we didn't have to remember to update this as well.

Reply:
Aside from a comment, does kube-builder/other code currently do enforcement or distinguish between various field support levels ?

Reply:
This was briefly discussed on the call today and we came to the conclusion that we do not currently validate different support levels. This is something that we will likely need for beta.

That said Rob mentioned a few options:

1. Keep this to the two mentioned filters
2. Revert to the original validation of "fail if there are more than 1 filter per rule"

Reply:
As I've thought about this more, I think the second option is probably the best here.

Reply:
Sounds good

Reply:
But doesn't that go against the whole idea of filters? We want to encourage users to use a chain of filters if their use-cases requires that.

Reply:
@robscott Can you respond to @hbagdi above ^

Reply:
Hey @cmluciano, sorry for the delay here! We discussed this at the community meeting on Wednesday to get a consensus. Harry made a compelling point that we don't know what future filter types could be and there could be generic ones that we want to allow repetition of. Because of that, your initial approach probably does make the most sense here. We know we don't want to allow repetition of the 2 filter types we currently have defined, but we don't want to restrict anything beyond that at this point.